📘 PERTEMUAN 5 - Support Vector Machine (SVM)

## Tujuan
1. Membandingkan kernel linear vs RBF pada dataset yang sama.
2. Melakukan tuning sederhana parameter C dan gamma.
3. Menentukan model terbaik.
4. Menuliskan 3 poin kesimpulan.

1️⃣ Perbandingan Kernel Linear vs RBF

Pada tahap pertama dilakukan perbandingan antara SVM dengan kernel linear dan RBF menggunakan dataset yang sama. 
Sebelum pelatihan, dilakukan standardisasi fitur karena SVM sensitif terhadap perbedaan skala.

In [1]:
# =========================================
# 1. PERBANDINGAN KERNEL
# =========================================

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report

# Load dataset
data = load_breast_cancer()

X_train, X_test, y_train, y_test = train_test_split(
    data.data, data.target, test_size=0.2, random_state=42
)

# Scaling
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Model Linear & RBF
svm_linear = SVC(kernel="linear", C=1)
svm_rbf = SVC(kernel="rbf", C=1, gamma="scale")

models = {
    "SVM Linear": svm_linear,
    "SVM RBF": svm_rbf
}

for name, model in models.items():
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    
    print("\n===================================")
    print(name)
    print("===================================")
    print("Accuracy :", round(accuracy_score(y_test, pred), 3))
    print(classification_report(
        y_test, pred, target_names=data.target_names
    ))


SVM Linear
Accuracy : 0.956
              precision    recall  f1-score   support

   malignant       0.93      0.95      0.94        43
      benign       0.97      0.96      0.96        71

    accuracy                           0.96       114
   macro avg       0.95      0.96      0.95       114
weighted avg       0.96      0.96      0.96       114


SVM RBF
Accuracy : 0.982
              precision    recall  f1-score   support

   malignant       1.00      0.95      0.98        43
      benign       0.97      1.00      0.99        71

    accuracy                           0.98       114
   macro avg       0.99      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114



## Tugas 2: Tuning Parameter C dan gamma

Parameter yang diuji:
- C = 0.1, 1, 10
- gamma = 0.01, 0.1, 1

Tuning dilakukan pada kernel RBF.

In [ ]:
# =========================================
# 2. TUNING C & GAMMA
# =========================================

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
import pandas as pd

# Load dataset
data = load_breast_cancer()

X_train, X_test, y_train, y_test = train_test_split(
    data.data, data.target, test_size=0.2, random_state=42
)

# Scaling
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

results = []

# Linear Kernel
for c in [0.1, 1, 10]:
    model = SVC(kernel="linear", C=c)
    model.fit(X_train, y_train)
    acc = accuracy_score(y_test, model.predict(X_test))
    results.append(["Linear", c, "-", round(acc,3)])

# RBF Kernel
for c in [0.1, 1, 10]:
    for gamma in [0.01, 0.1, 1]:
        model = SVC(kernel="rbf", C=c, gamma=gamma)
        model.fit(X_train, y_train)
        acc = accuracy_score(y_test, model.predict(X_test))
        results.append(["RBF", c, gamma, round(acc,3)])

# Buat tabel
df_results = pd.DataFrame(results, columns=["Kernel", "C", "Gamma", "Accuracy"])

print("\n===================================")
print("HASIL TUNING PARAMETER")
print("===================================")
print(df_results)

# Ambil model terbaik
best = df_results.loc[df_results["Accuracy"].idxmax()]

print("\n===================================")
print(" MODEL TERBAIK")
print("===================================")
print(best)


HASIL TUNING PARAMETER
    Kernel     C Gamma  Accuracy
0   Linear   0.1     -     0.982
1   Linear   1.0     -     0.956
2   Linear  10.0     -     0.965
3      RBF   0.1  0.01     0.965
4      RBF   0.1   0.1     0.947
5      RBF   0.1     1     0.623
6      RBF   1.0  0.01     0.965
7      RBF   1.0   0.1     0.965
8      RBF   1.0     1     0.632
9      RBF  10.0  0.01     0.982
10     RBF  10.0   0.1     0.965
11     RBF  10.0     1     0.632

 MODEL TERBAIK
Kernel      Linear
C              0.1
Gamma            -
Accuracy     0.982
Name: 0, dtype: object


3️⃣ Kesimpulan

1️. Kernel Linear dengan C = 0.1 menghasilkan accuracy tertinggi (0.982), sehingga menjadi model terbaik pada percobaan ini. Hal ini menunjukkan bahwa data relatif dapat dipisahkan secara linear.

2️. Parameter C mempengaruhi keseimbangan antara margin dan toleransi error. Nilai C yang lebih kecil (0.1) justru memberikan performa lebih baik dibanding C yang lebih besar, karena menghasilkan margin yang lebih stabil dan mengurangi risiko overfitting.

3️. Pada kernel RBF, nilai gamma yang terlalu besar (gamma = 1) menyebabkan accuracy turun drastis. Hal ini menunjukkan bahwa gamma tinggi membuat batas keputusan terlalu kompleks dan berisiko overfitting.